atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionCacheValueStoreImpl.java

```diff
@@ -0,0 +1,154 @@
+                    .filter(snapshot::isUnlocked)
+                    .forEach(cell -> localUpdates.put(cell, LocalCacheEntry.read(CacheValue.empty())));
+        }
+    }
+
+    @Override
+    public Map<CellReference, CacheValue> getCachedValues(Set<CellReference> cellReferences) {
+        Map<CellReference, CacheValue> locallyCachedValues = getLocallyCachedValues(cellReferences);
+
+        // Filter out which values have not been read yet
+        Set<CellReference> remainingCells = Sets.difference(cellReferences, locallyCachedValues.keySet());
+
+        // Read values from the snapshot. For the hits, mark as hit in the local map.
+        Map<CellReference, CacheValue> snapshotCachedValues = getSnapshotValues(remainingCells);
+        snapshotCachedValues.forEach(
+                (cellReference, value) -> localUpdates.put(cellReference, LocalCacheEntry.hit(value)));
+
+        return ImmutableMap.<CellReference, CacheValue>builder()
+                .putAll(locallyCachedValues)
+                .putAll(snapshotCachedValues)
+                .build();
+    }
+
+    @Override
+    public Map<CellReference, CacheValue> getTransactionDigest() {
+        return KeyedStream.stream(localUpdates)
+                .filter(entry -> entry.status().equals(Status.READ))
+                .map(LocalCacheEntry::value)
+                .collectToMap();
+    }
+
+    private Map<CellReference, CacheValue> getLocallyCachedValues(Set<CellReference> cellReferences) {
+        return KeyedStream.of(cellReferences)
+                .map(localUpdates::get)
+                .filter(Objects::nonNull)
+                .map(LocalCacheEntry::value)
+                .collectToMap();
+    }
+
+    private Map<CellReference, CacheValue> getSnapshotValues(Set<CellReference> cellReferences) {
+        return KeyedStream.of(cellReferences)
+                .map(snapshot::getValue)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .filter(CacheEntry::isUnlocked)
+                .map(CacheEntry::value)
+                .collectToMap();
+    }
+
+    @Value.Immutable
+    public interface LocalCacheEntry {
+        Status status();
+
+        CacheValue value();
+
+        static LocalCacheEntry read(CacheValue value) {
+            return ImmutableLocalCacheEntry.builder()
+                    .status(Status.READ)
+                    .value(value)
+                    .build();
+        }
+
+        static LocalCacheEntry write(CacheValue value) {
+            return ImmutableLocalCacheEntry.builder()
+                    .status(Status.WRITE)
+                    .value(value)
+                    .build();
+        }
+
+        static LocalCacheEntry hit(CacheValue value) {
+            return ImmutableLocalCacheEntry.builder()
+                    .status(Status.HIT)
+                    .value(value)
+                    .build();
+        }
+
+        enum Status {
+            READ,
+            WRITE,
```

Comment:
The distinction between WRITE and HIT here is just for cleaner code, we only really care about whether it was a remote read or not?

Reply:
No - the distinction is for later, when we only want `HIT` values for conflict checking (we don't need to do serialisable conflict checking for `READ` values as those are read from the DB, whereas we do need to check the `HIT` values as they have not touched the conflict checking flow at all).

Reply:
👍